<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-файла-с-данными-и-общей-информации" data-toc-modified-id="Изучение-файла-с-данными-и-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение файла с данными и общей информации</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></div>

# Описание проекта

    Заказчик - компания крупного сервиса для чтения книг по подписке
    Задача — проанализировать базу данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг.


# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

##  Изучение файла с данными и общей информации

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
              'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
               'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [5]:
# выведем первые пять строк таблиц 
for query in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    request=pd.io.sql.read_sql(query, con = engine)
    display('Таблица {}:'.format(query), request.head(),  request.info(),'='*90)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   book_id           1000 non-null   int64         
 1   author_id         1000 non-null   int64         
 2   title             1000 non-null   object        
 3   num_pages         1000 non-null   int64         
 4   publication_date  1000 non-null   datetime64[ns]
 5   publisher_id      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


'Таблица books:'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

None

'=========================================================================================='

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


'Таблица authors:'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

None

'=========================================================================================='

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


'Таблица publishers:'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

None

'=========================================================================================='

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


'Таблица ratings:'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

None

'=========================================================================================='

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


'Таблица reviews:'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

None

'=========================================================================================='

In [6]:
# напишем функцию для упрощения вывода query запроса  
def table(query):
    ('''   ''')
    return pd.io.sql.read_sql(query, con = engine)

# Описание данных

В распоряжении датасеты с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

- таблица books - 1000 строк, 6 столбцов с типом данных datetime64[ns](1), int64(4), object(1)
- таблица authors - 636 строк, 2 столбца с типом данных int64(1), object(1)
- таблица publishers - 340 строк, 2 столбца с типом данных int64(1), object(1)
- таблица ratings - 6456 строки, 4 столбца с типом данных int64(3), object(1)
- таблица reviews - 2793 строки, 4 столбца с типом данных int64(2), object(2)


**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Посчитайте, сколько книг вышло после 1 января 2000 года

- определяю для вывода на экран столбцы с посчетом количества книг book_id
- описываем откуда выгружаем данные - таблица с данными о книгах books
- с помощью оператора сравнения WHERE отбираю книги по нужной дате publication_date

In [5]:
table('''
SELECT 
    COUNT(book_id) AS number_book
FROM 
    books
WHERE 
    publication_date > '2000-01-01'
''')

number_book
0          819

после 1 января 2000 года вышло 819 книг

## Для каждой книги посчитайте количество обзоров и среднюю оценку

- определяю для вывода на экран столбцы с идентификатором книги book_id, названием книги title, считаю уникальное количество по  идентификатору обзора review_id, вычисляю среднюю оценку по оценке книги rating
- описываем откуда выгружаем данные - таблица books
- объединяю по таблице с данными о книгах books таблицы с данными о пользовательских оценках книг ratings и обзорах книг reviews по идентификатору книги book_id
- группирую данные по идентификатор книги book_id
- сортирую количество оценок по убыванию number_review

In [7]:
table('''
SELECT
    b.book_id,
    b.title,
    COUNT(DISTINCT rv.review_id) AS number_review,
    AVG(rt.rating) AS mean_rating
FROM
    books AS b
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
GROUP BY
    b.book_id
ORDER BY
    number_review DESC
''')

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     number_review  mean_rating  
0                7     3.662500  
1                6     3.977273  
2                6     4.206897  
3                6     4.414634  
4                6     4.081081  
..             ...          ...  
995              0     3.666667  
996              0     3.000000  
997              0     5.000000  
998              0     4.000000  
999              0     4.000000  

[1000 rows x 4 columns]

всего получилось 1000 книг, с количеством обзоров от 0 до 7, для каждой книги выведена средняя оценка

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

Брошю́ра — печатное произведение небольшого объёма и ее размер не менее 5 и не более 48 страниц, поэтому отберем в столбце количество страниц books.num_pages, только те книги, у которых страниц больше 50

- определяю для вывода на экран наименование издательства publisher и считаю количество книг по идентификатор книги book_id
- описываем откуда выгружаем данные - таблица books
- по таблице с данными о книгах books присоединяю данные об издательствах publishers по идентификатор издательства publisher_id
- с помощью оператора сравнения WHERE отбираю страницы num_pages, количество которых больше 50
- делаю группировку GROUP BY по идентификатору издательства и наименованию издательства
- сортирую ORDER BY по количеству отобранных книг number_book по убыванию
- вывожу только первую строку LIMIT

In [7]:
table('''
SELECT
    p.publisher,
    COUNT(b.book_id) AS number_book
FROM
    books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY
        b.publisher_id,
        p.publisher
ORDER BY 
       number_book DESC 
LIMIT 1
''')

publisher  number_book
0  Penguin Books           42

Издательство выпустившее наибольшее число книг исключая брощюры стало Penguin Books - 42 книги 

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

- определяю для вывода на экран столбцы с автором author и вычисляю средний рейтинг rating с помощью AVG
- описываем откуда выгружаем данные - таблица books
- объединяю по таблице books, таблицу authors по идентификатору автора author_id и таблицу ratings по идентификатор книги book_id
- группировка по идентификатору автора author_id и book_id
- после группировки с помощью оператора HAVING делаю срез данных по количеству rating_id идентификатор оценки равную и больше 50 оценок
- сортирую с помощью оператора ORDER BY средний рейтинг mean_rating по убыванию DESC
- вывожу первые первую строку с помощью оператора LIMIT

In [10]:
table('''
SELECT 
    a.author,
    AVG(r.rating) AS mean_rating
FROM (SELECT book_id,
    AVG(rating) AS rating_avg
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50) AS rt
JOIN ratings AS r on rt.book_id=r.book_id 
JOIN books AS b ON rt.book_id = b.book_id
JOIN authors AS a ON b.author_id = a.author_id

GROUP BY
        a.author       
ORDER BY mean_rating DESC
LIMIT 1
''')

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.287097

самый высокий рейтинг с оценкой равной или более 50 стали книги написанные совместно авторами J.K. Rowling/Mary GrandPré с рейтингом 4.287097

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

- определяю для вывода на экран среднее количество оценкок вычисляем: количество обзоров по идентификатору  обзора делим на уникальное количество пользователей написавших обзор
- описываем откуда выгружаем данные - таблица reviews 
- с помощью оператора сравнения WHERE делаю вложенный запрос для пользователей, где отбираю пользовтаелй в таблице с рейтингом по идентификатору оценок с количеством оценок больше 50

In [9]:
table('''
SELECT 
    COUNT(review_id)/ COUNT(DISTINCT(username)) AS mean_review 
FROM
    reviews
WHERE  username IN 
   (SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
)
''')

mean_review
0           24

в среднем на каждого пользователя приходятся по 24 отзыва

# Выводы

- после 1 января 2000 года вышло 819 книг
- всего получилось 1000 книг, с количеством обзоров от 0 до 7, для каждой книги выведена средняя оценка
- издательство выпустившее наибольшее число книг исключая брощюры стало Penguin Books - 42 книги 
- самый высокий рейтинг с оценкой равной или более 50 стали книги написанные совместно авторами J.K. Rowling/Mary GrandPré с рейтингом 4.287097
- в среднем на каждого пользователя приходятся по 24 отзыва, которые поставили больше 50 оценок